oak/server/rust/oak_runtime/src/node/wasm.rs

```diff
@@ -86,21 +86,18 @@ struct WasmInterface {
 impl WasmInterface {
     /// Generate a randomized handle. Handles are random to prevent accidental dependency on
     /// particular values or sequence. See https://github.com/project-oak/oak/pull/347
-    fn allocate_new_handle(&mut self, channel: ChannelEither) -> AbiHandle {
+    fn allocate_new_handle(&mut self, channel: Handle, direction: HandleDirection) -> AbiHandle {
```

Comment:
I'm still confused by `Handle` vs `AbiHandle`. Why do they both need to exist, and are both randomized? We should be able to just use the underlying `Handle` here?

Reply:
Apart from `oak_abi::Handle` there is
- `channel::Handle` which is `pub struct Handle(oak_abi::Handle);`, I was thinking it might be useful to have this completely opaque, but since the handles are randomized maybe there isn't much reason WDYT?
- `Wasm::AbiHandle` is `type AbiHandle = u64` and will be removed/replaced with oak_abi::handle. I was holding off on changing this until the next PR with tracking channels in nodes, so that I can remove all of the "handle" generation from Wasm and only use the runtime handles (and also enforce that nodes only access handles they own).

Reply:
> Apart from `oak_abi::Handle` there is
> 
> * `channel::Handle` which is `pub struct Handle(oak_abi::Handle);`, I was thinking it might be useful to have this completely opaque, but since the handles are randomized maybe there isn't much reason WDYT?

That's a good question. On one hand, it would be nice for this to be type-safe in Rust (modulo unsafe code), in the sense of not being forgeable out of thin air. On the other hand, a Wasm node does rely on the fact that the Wasm module it is executing will provide handles as numbers, and the runtime around the Wasm interpreter will turn those numbers into Handles in order to pass them to the runtime.

How about the following proposal:

- ids (e.g. node_id, channel_id) are generated via incrementing counters, but are tracked and protected by the Rust type system (i.e. unforgeable, even if someone guessed the underlying number, they would not be able to wrap it in the correct type). We have to assume this at least for node_id, since if we are not trusting the Rust type system to provide this safety, then all bets are off (a malicious node may just impersonate another node when interacting with the runtime).
- handles only exist in the context of a Wasm node instance, and the Wasm node itself generates them randomly (**not** the Runtime), and keeps track of the mapping from each random / unguessable handle to a corresponding typesafe id object locally.

In this model, the `channel_read` ABI call would be defined in terms of ids (caller node id, target channel id) on the Runtime itself, but the Wasm node would instead:

- not expose the caller node id (since it is implicitly bound to the current node instance, which is enforced by a RuntimeProxy object or similar in the type system)
- accept a channel handle as parameter, which is then looked up by the Wasm node to figure out the underlying channel id (if any), which is then passed to the underlying call on the RuntimeProxy object. The Runtime never sees the handle, only the id.

WDYT?

@daviddrysdale thoughts?

> * `Wasm::AbiHandle` is `type AbiHandle = u64` and will be removed/replaced with oak_abi::handle. I was holding off on changing this until the next PR with tracking channels in nodes, so that I can remove all of the "handle" generation from Wasm and only use the runtime handles (and also enforce that nodes only access handles they own).

Reply:
> WDYT?

I like this idea, although pseudo nodes not having handles means that tracking when channels become orphaned is more complex.

Regarding this PR, WDYT can I merge and then work on tracking nodes with channels in a follow up PR ?